### PR TITLE
gql2go,graphql: Add support for recursive input types via init loading

### PR DESCRIPTION
**Pull request by michaelraines**
Example use case

```
input WorkflowNodeInput {
  name: String!
  action: WorkflowNodeActionInput!
  children: [WorkflowNodeRelationshipInput]!
  triggers: [WorkflowNodeTriggerInput!]!
}

input WorkflowNodeRelationshipInput {
  child: WorkflowNodeInput!
  continuationType: WorkflowNodeRelationshipContinuationType!
}
```

### CODE REVIEW

File: definition.go [hunk 1]
```diff
@@ -1143,6 +1147,18 @@ func NewInputObject(config InputObjectConfig) *InputObject {
 	gt.fields = gt.defineFieldMap()
 	return gt
 }
+func (st *InputObject) AddInputField(fieldName string, fieldConfig *InputObjectField) {
+	if fieldName == "" || fieldConfig == nil {
+		return
+	}
+	st.mu.Lock()
```
Comment (samuel) [inline definition.go]: Need to check (R)Lock this mutex anywhere that st.fields or st.typeConfig.Fields are accessed (e.g. `(*InputObject).Fields()`)

Comment (michaelraines) [reply definition.go]: Good call. Fixed.

File: definition.go [hunk 2]
```diff
@@ -1180,6 +1198,8 @@ func (gt *InputObject) defineFieldMap() InputObjectFieldMap {
 	return resultFieldMap
 }
 func (gt *InputObject) Fields() InputObjectFieldMap {
+	gt.mu.Lock()
```
Comment (samuel) [inline definition.go:1209]: This functions I think needs to match the one for *Object. That is, if gt.fields is nil then it needs to be rebuilt as AddInputField nils it out.

https://github.com/sprucehealth/graphql/blob/c9b790deef52145485f35e397e7f4c18f49cb75d/definition.go#L445

Comment (michaelraines) [reply definition.go:1209]: Good call. I think input object handles this in (*InputObject).defineFieldMap and just needs to be recalled if it's nil after an update. Made that change.

**Review by samuel — APPROVED** [review]

**Pull request merged**